_posts/2018-08-26-eager-dcgan/eager-dcgan.Rmd

```diff
@@ -0,0 +1,460 @@
+
+In human learning, feedback - direct or indirect - plays a central role. Say we wanted to forge a banknote (as long as those still exist). Assuming we can get away with unsuccessful trials, we would learn from every rejection and distrustful inspection. Optimizing our technique, we would end up rich. 
+This concept of optimizing from feedback is embodied in the first of the two agents, the _generator_. It gets its feedback from the _discriminator_, in an upside-down way: If it can fool the discriminator, making it believe that the banknote was real, all is fine; if the discriminator notices the fake, it has to do things differently. For a neural network, that means it has to update its weights.
+
+How does the discriminator know what is real and what is fake? It too has to be trained, on real banknotes (or whatever the kind of objects involved) and the fake ones produced by the generator. So the complete setup is two agents competing, one striving to generate realistic-looking fake objects, and the other, to disavow the deception. The purpose of training is to have both evolve and get better, in turn causing the other to get better, too.
+
+In this system, there is no objective minimum to the loss function: We want both components to learn and getter better "in lockstep", instead of one winning out over the other. This makes optimization difficult.
+In practice therefore, tuning a GAN can seem more like alchemy than like science, and it often makes sense to lean on practices and "tricks" reported by others.
+
+In this example, just like in the Google notebook we're porting, the goal is to generate MNIST digits. While that may not sound like the most exciting task one could imagine, it lets us focus on the mechanics, and allows us to keep computation and memory requirements (comparatively) low.
+
+Let's load the data (training set needed only) and then, look at the first actor in our drama, the generator.
+
+## Training data
+
+```{r}
+mnist <- dataset_mnist()
+c(train_images, train_labels) %<-% mnist$train
+
+train_images <- train_images %>% 
+  k_expand_dims() %>%
+  k_cast(dtype = "float32")
+
+train_images <- (train_images - 127.5) / 127.5
+```
+
+Our complete training set will be streamed once per epoch:
+
+```{r}
+buffer_size <- 60000
+batch_size <- 256L
+batches_per_epoch <- (buffer_size / batch_size) %>% round()
+
+train_dataset <- tensor_slices_dataset(train_images) %>%
+  dataset_shuffle(buffer_size) %>%
+  dataset_batch(batch_size)
+```
+
+This input will be fed to the discriminator only.
+
+## Generator
+
+Both generator and discriminator are [custom models](https://keras.rstudio.com/articles/custom_models.html).
+As we will soon see, the generator gets passed vectors of random noise for input. This vector is transformed to 3d (height, width, channels) and then, successively upsampled to the required output size of (28,28,3).
+
+
+```{r}
```

Comment:
Maybe just briefly reference what a custom model is (custom forward pass logic) and how it works (initializes and assigns layers to self and returns the function that implements the forward pass)

Reply:
suggesting

```
In contrast to custom layers, custom models allow you to construct models as independent units, complete with custom forward pass logic, backprop and optimization. The model-generating function defines the layers the model (`self`) wants assigned, and returns the function that implements the forward pass.
```